Stats Insights: Optimize table view updates by integrating diffable data source

staskus opened the pull request:
Use a diffable data source within Stats Insights and Stats Insights Details to make table view cell reloads more efficient. More information about implementation in: https://github.com/wordpress-mobile/WordPress-iOS/pull/22542

Related diff: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/751
 
## To test:

1. Open Stats Insights
2. Confirm data loads correctly without flickering
3. Tap on Insights management button on top right
4. Remove all cards
5. Confirm empty view with a button is shown
6. Add all cards
7. Tap "View More" on all the cards
8. Confirm Stats Insights Details view opens

## Regression Notes
1. Potential unintended areas of impact

Breaking loading of data within existing Stats Insights and Stats Insights Details. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- Manual testing
- `SiteStatsInsightsViewModelTests` should still succeed

9. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

staskus committed:
Prepare StatsInsights for diffable by making entities Hashable

staskus committed:
Update SiteStatsInsightsViewModel to build diffable data source snapshot

staskus committed:
Apply snapshot instead of reloading table in SiteStatsInsightsTableViewController

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Integrate ImmuTableDiffableViewHandler

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

staskus commented:
@guarani review for this can wait after we're done with Traffic tab.

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

staskus committed:
Make Stats entities Equatable in WordPressKit

staskus committed:
Rename StatsTrafficModels to HashableImmutableRow

staskus committed:
Make Stats Insights Details compatible with diffable data source

staskus commented:
I added Stats Insights Details support in this PR. 
Stats Period Details is implemented in https://github.com/wordpress-mobile/WordPress-iOS/pull/22823.

These two PRs should cover diffable data source support for all the Stats.

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

guarani commented:
Related to the failing CI checks: https://a8c.slack.com/archives/C05N140C8H5/p1710357494940709

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

guarani commented:
I moved this to a new GitHub project, more context here: p1710464002701159-slack-C06BR07TJHK

staskus commented:
TODO: Requires the same fix https://github.com/wordpress-mobile/WordPress-iOS/pull/22857

Done:

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

staskus committed:
Update StatsInsights VC when toggling using diffable data source apply methods

Usage of performBatchUpdates is not allowed if table view depends on a diffable data source and can result in crashes

staskus committed:
Fix compile errors after updating WPKit

staskus commented on WordPress/Classes/Services/PostRepository+Helpers.swift:
```diff
@@ -4,7 +4,10 @@ import WordPressKit
     /// Initializes the parameters required to create the given post.
     init(post: AbstractPost) {
-        self.init(status: (post.status ?? .draft).rawValue)
+        self.init(
```
Related to this change https://github.com/wordpress-mobile/WordPressKit-iOS/pull/757, I copied the solution from this PR https://github.com/wordpress-mobile/WordPress-iOS/pull/22848

staskus replied:
The CI continues to fail, it will be problematic until all the changes in https://github.com/wordpress-mobile/WordPress-iOS/pull/22849 are merged. 

Lets review this with failing CI, once we need to merge, we can coordinate with the other work to get this to pass.

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

guarani reviewed (approved):
I tested this and can see the improvement!
The changes themselves were hard to review due to the number of lines updated, so I skimmed the changes and relied on testing.

staskus committed:
Merge branch 'trunk' into task/stats-insights-diffable-optimization

staskus committed:
Update WordPressKit version

staskus committed:
Update RELEASE-NOTES.txt
